### PR TITLE
Standardize scalar-valued external function interfaces

### LINKED ISSUE #2157: Add `evaluate_fgh` to PythonCallbackFunction API
## Summary

ExternalFunctions through ASL offer calls to `evaluate_fgh` to provide derivative information. We should add this feature to the `PythonCallbackFunction` API as well.

### Rationale

The APIs should match for consistency across different types of ExternalFunctions.

### PR DESCRIPTION
**Pull request by jsiirola**
## Summary/Motivation:
The Python and ASL external function interfaces have drifted apart over time.  When we added `evaluate_fgh` to the ASL interface, we did not also add it to the Python callback interface.  This PR reworks the Scalar-values ExternalFunction interface so that `evaluate()` and `evaluate_fgh()` are cleanly supported through both ASL and PythonCallback interfaces.

In addition, there are numerous code cleanups:
- G, and H entries are now zeroed out based on the fixed flag (i.e., the gradient and Hessian values are now guaranteed to be correct when specifying that argument indices are fixed).
- improved handling of the global "function id" in the PythonCallbackFunction interface, including special handling so that `clone()` / `deepcopy()` work correctly.
- Improved documentation

## Changes proposed in this PR:
- Add `evaluate_fgh` as a formal part of the `ExternalFunction` API
- Resolve pickling issues with the PythonCallbackFunction
- Improve test coverage
- Add documentation

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/core/base/external.py [hunk 1]
```diff
@@ -28,27 +31,103 @@
 
 
 class ExternalFunction(Component):
+    """Interface to an external (non-algrbraic) function.
```
Comment (mrmundt) [inline pyomo/core/base/external.py:34]: Typo: `algebraic`

File: pyomo/core/base/external.py [hunk 2]
```diff
@@ -28,27 +31,103 @@
...
 
-    def __new__(cls, *args, **kwds):
-        if cls != ExternalFunction:
+    :class:`ExternalFunction` provides an interface for declaring
+    general user-provided functiona, and then embedding calls to the
```
Comment (mrmundt) [inline pyomo/core/base/external.py:37]: Typo: `functions`

File: pyomo/core/base/external.py [hunk 3]
```diff
@@ -28,27 +31,103 @@
...
+    def __new__(cls, *args, **kwargs):
+        if cls is not ExternalFunction:
             return super(ExternalFunction, cls).__new__(cls)
-        if len(args) == 1 and type(args[0]) is types.FunctionType:
-            return PythonCallbackFunction.__new__(PythonCallbackFunction)
-        elif not args and 'library' not in kwds and 'function' in kwds and \
-                type(kwds['function']) is types.FunctionType:
-            return PythonCallbackFunction.__new__(PythonCallbackFunction)
+        elif args:
+            return super(ExternalFunction, cls).__new__(PythonCallbackFunction)
```
Comment (mrmundt) [inline pyomo/core/base/external.py:55]: I'm going to recommend that this be changed to follow Python3 standards:
```
return super().__new__(PythonCallbackFunction)
```

File: pyomo/core/base/external.py [hunk 4]
```diff
@@ -28,27 +31,103 @@
...
+        elif 'library' not in kwargs and any(
+                kw in kwargs for kw in ('function', 'fgh')):
+            return super(ExternalFunction, cls).__new__(PythonCallbackFunction)
```
Comment (mrmundt) [inline pyomo/core/base/external.py:58]: I'm going to recommend that this be changed to follow Python3 standards:
```
return super().__new__(PythonCallbackFunction)
```

File: pyomo/core/base/external.py [hunk 5]
```diff
@@ -28,27 +31,103 @@
...
+
+          A single external function call with a signature matching the
+          :meth:`evaluate_fgh()` method.
+
+        2. Independent functions:
+
+          One to three functions that can evaluate the function value,
+          gradient of the function [partial derivatives] with respect to
+          its inputs, and the Hessian of the function [partial second
+          dericatives].  The ``function`` interface expects a function
```
Comment (mrmundt) [inline pyomo/core/base/external.py:91]: Typo: `derivatives`

File: pyomo/core/base/external.py [hunk 6]
```diff
@@ -28,27 +31,103 @@
...
+          matching the prototype:
+
+          .. code::
+
+             def function(*args): float
+
+          The ``gradient`` amd ``hessian`` interface expect functions
```
Comment (mrmundt) [inline pyomo/core/base/external.py:98]: Typo: `and`

File: pyomo/core/base/external.py [hunk 7]
```diff
@@ -201,59 +407,151 @@ def _pprint(self):
         )
 
 
+class _PythonCallbackFunctionID(NumericConstant):
+    """A specialized NumericConstant to preserve FunctionIDs through deepcopy.
+
+    Function IDs are effectively pointers back to the
+    PythonCallbackFunction.  As such, we need special handling to
+    maintain / preserve the correct linkages through deepcopy (and
+    model.clone().
```
Comment (mrmundt) [inline pyomo/core/base/external.py:416]: Missing close parenthesis here.

File: pyomo/core/base/external.py [hunk 8]
```diff
@@ -201,59 +407,151 @@ def _pprint(self):
...
+            if fgh == 2:
+                if self._hess is None:
+                    raise RuntimeError(
+                        f"ExternalFunction '{self.name}' was not defined "
+                        "with a Hessian callback.  Cannot evaluate the "
+                        "second derivative of the function")
+                h = self._hess(args, fixed)
+            else:
+                h = None
+        # Update g,h to reflect the function id (which by defiition is
```
Comment (mrmundt) [inline pyomo/core/base/external.py:542]: Typo: `definition`

**Review by mrmundt — CHANGES_REQUESTED** [review]

File: pyomo/core/base/external.py [hunk 9]
```diff
@@ -201,59 +407,151 @@ def _pprint(self):
...
     def register_instance(cls, instance):
-        _id = len(cls.global_registry)
-        cls.global_registry[_id] = weakref.ref(instance)
-        return _id
-
-    def __init__(self, *args, **kwds):
-        if args:
-            self._fcn = args[0]
-            if len(args) > 1:
+        # Ideally, we would just use the id() as the finction id, but
```
Comment (blnicho) [inline pyomo/core/base/external.py:444]: Typo: function

**Review by blnicho — APPROVED** [review]
I think this looks fine, just a few typos to fix.

**Review by mrmundt — APPROVED** [review]

**Pull request merged**